vault/identity_store_entities.go

```diff
@@ -799,23 +861,41 @@ func (i *IdentityStore) mergeEntity(ctx context.Context, txn *memdb.Txn, toEntit
 			return errors.New("entity id to merge from does not belong to this namespace"), nil
 		}
 
-		for _, alias := range fromEntity.Aliases {
+		for _, fromAlias := range fromEntity.Aliases {
+			if toAliasId, ok := toEntityAccessors[fromAlias.MountAccessor]; ok {
+				// Handle conflicts (conflict = both aliases share the same mount accessor)
+				if strutil.StrListContains(conflictingAliasIDsToKeep, toAliasId) {
+					i.logger.Info("Deleting from_entity alias during entity merge", "from_entity", fromEntityID, "deleted_alias", fromAlias.ID)
+					err := i.MemDBDeleteAliasByIDInTxn(txn, fromAlias.ID, false)
+					if err != nil {
+						return nil, fmt.Errorf("failed to delete orphaned alias during merge: %w", err)
+					}
+
+					// Continue to next alias, as there's no alias to merge left in the from_entity
+					continue
+				} else if forceMergeAliases || strutil.StrListContains(conflictingAliasIDsToKeep, fromAlias.ID) {
+					i.logger.Info("Deleting to_entity alias during entity merge", "to_entity", toEntity.ID, "deleted_alias", toAliasId)
+					err := i.MemDBDeleteAliasByIDInTxn(txn, toAliasId, false)
+					if err != nil {
+						return nil, fmt.Errorf("failed to delete orphaned alias during merge: %w", err)
+					}
+				} else {
+					return fmt.Errorf("conflicting mount accessors in following alias IDs and neither were present in conflicting_alias_ids_to_keep: %s, %s", fromAlias.ID, toAliasId), nil
```

Comment:
I'm worried about this line, because if there's an alias that's somehow snuck into an entity and on login a merge is required, and that alias isn't the alias used to login via aliasByFactors, then the entity will no longer be usable.

Reply:
This path won't be used in the automated merge case - the previous condition check includes a check for if `forceMergeAliases` is true. This will only occur in the manual case where the user hasn't provided the required aliasIds to perform a merge.

Reply:
I'm not sure I follow. Which condition check are you referring to?

Reply:
Ah, I see now. The else-if will always catch the automated merge case because of `forceMergeAliases` . Makes sense to me, but I still don't understand why this else block exists, and why the automatic case and the manual case are being checked in the same for loop.

Reply:
The else block exists if there are clashing aliases that haven't been added to the `conflictingAliasIDsToKeep` array. For example, maybe there are two clashes, and the customer only provides an aliasId for one clash.